README.md
# TrackioUno - GPS/GPRS Tracker and IoT Device

TrackioUno es un firmware escrito en C++ especialmente diseñado para la familia SAMD21 (Cortex-M0) de Microchip. Entre sus principales funcionalidades destaca la posibilidad de trabajar manteniendo un TCP permanentemente abierto, lo que permite una comunicación bidireccional completa entre cliente-servidor.

En Trackio.h se ha documentado cada método de la clase Trackio y hay muchas referencias sobre el uso del módulo Sim868 de Rhomb.io, por lo que también puede ser útil para aprender a utilizarlo.

El firmware está compuesto básicamente de 3 archivos, la clase Trackio en los ficheros Trackio.h y Trackio.cpp y el archivo  main.cpp. Se incluye además el archivo _rhio-pins.h_ para la configuración de pins con la plataforma Rhomb.io y _static-conf.h_ para la configuración de algunos parámetros estáticos no modificables en tiempo de ejecución. También se han incluído algunas dependencias de terceros para gestión de Watchdog/Sleep y FlashMemory

## Descargar el Código

A partir de de la versión 0.3.0 se da compapatibilidad a Duino Zero. Para Duino Uno/Mega se debe usar una versión 0.2.x

TrackioUno está disponible en [Github](https://github.com/M2MSystemSource/TrackioUno. Se ofrecen distintas ramas:

* [master](#) Última versión estable (sin bugs conocidos)
* [develop](#) Versión en desarrollo (NO estable)

 Puedes descargar la [rama master](https://github.com/M2MSystemSource/TrackioUno/archive/master.zip) con los últimos cambios aunque puede no ser estable, si necesitas una versión segura es mejor descargar una [release](https://github.com/M2MSystemSource/TrackioUno/releases).

**IMPORTANTE!** Debes duplicar los archivos `platformio.ini.sample` y `lib/Trackio/static-conf.h.sample` eliminando la extensión `.sample` del modo que obtendrás `platformio.ini` y `lib/Trackio/static-conf.h`. **Si piensas publicar cambios en Git no elimines los archivos `.sample` para que no desaparezcan del repositorio.**

### Duino Uno

La anterior versión, v0.2.x es compatible con Duino Uno (atmega328p) y Mega (Atmega2560/1280). Se mantiene la rama [DuinoUno](https://github.com/M2MSystemSource/TrackioUno/tree/DuinoUno) con esta versión, ademas de todas las releases de esta rama.

## IDE de programación y upload

TrackioUno ha sido creado con [PlatformIO](https://platformio.org/) sobre VS Code. PlatformIO es una utilidad de consola ([PlatformIO Core](http://docs.platformio.org/en/latest/core.html)) que además tiene el plugin ([PlatformIO IDE](https://platformio.org/platformio-ide)) para los editores [Atom](https://atom.io) de Github y [VS Code](https://code.visualstudio.com/) de Microsoft, ambos multiplataforma (Windows/Linux/Mac) y Open Source.

PlatformIO dispone de guías muy completas para su instalación tanto en Atom como en VS Code. El que suscribe ha tenido una mejor experiencia con VS Code.

- [Instalación de PlatfomIO en Atom](http://docs.platformio.org/en/latest/ide/atom.html)
- [Instalación de PlatfomIO en VS Code](http://docs.platformio.org/en/latest/ide/vscode.html)

Al ser una aplicación de consola no es obligatorio usar uno de estos editores, puedes gastar cualquier otro y seguir los pasos en la documentación de Platformio para hacer los flasheos.

Resumiendo la instalación, primero tendremos que instalar el editor, entramos a su web y descargamos la versión para nuestro sistema operativo. En windows tendremos un instalador (siguiente, siguiente... finalizar) y para linux un archivo _.dev_ o _.rpm_. Seguimos los pasos de instalación en cualquier caso. La descarga es completamente gratuita y sin registro.

Una vez instalado el editor debemos utilizar su gestor de plugins, que se conecta a internet para obtener un listado de los plugins disponibles y entre ellos encontraremos PlatformIO (utiliza el buscador). Al instalar PlatformIO veremos su pantalla de Home, donde podemos crear un nuevo proyecto. La primera vez que crees un proyecto PlatformIO tendrá que descargar todas sus dependencias, en el caso de los master _Duino_ de _Rhomb.io_ esto incluye el framework de _Arduino_ (Arduino Core) y las librerías de _Avrlib_ (_avr-tools_, _avrdude_...) por lo que tardará un tiempo en iniciar el proyecto. Los siguientes proyectos se crearán de forma automática sin esperas.

> Si utilizas Directamente TrackioUno sin haber creado previamente un proyecto propio, cuando compiles y hagas el primer upload PlatformIO detectará que no tiene las dependencias por lo que empezará a descargarlas y verás el proceso en la consola. Al acabar la descarga continuará con el upload del código.

## Soporte para módulos master

Desde la version 0.3.x el soporte de módulos master se limita a Duino Zero. La serie 0.2.x Soporta los mcus con arquitectura AVR, Duino Uno y Mega

### Configuración de platformio.ini para upload

En la raíz del proyecto deberemos haber creado el archivo _platformio.ini_ a partir de _platformio.ini.sample_, donde le indicamos a PlatformIO que compilador y método de upload debe utilizar.

~~~ini
[env:zeroUSB]
platform = atmelsam
board = zeroUSB
framework = arduino
build_flags = -D SERIAL_BUFFER_SIZE=128
upload_port = /dev/ttyACM0
~~~

El punto y coma indica un comentario, por lo que ahora no estaría disponible ninguna opción, para programar un _Duino Uno_ descomentas sus variables al tiempo que el resto quedan comentadas:

~~~ini
[env:uno]
platform = atmelavr
board = uno
framework = arduino
upload_port = /dev/ttyUSB0
build_flags = -D SERIAL_RX_BUFFER_SIZE=128

;[env:megaatmega2560]
;platform = atmelavr
;board = megaatmega2560
;framework = arduino
;build_flags = -D SERIAL_RX_BUFFER_SIZE=128
;upload_port = /dev/ttyUSB0

;[env:zeroUSB]
;platform = atmelsam
;board = zeroUSB
;framework = arduino
;build_flags = -D SERIAL_BUFFER_SIZE=128
;upload_port = /dev/ttyACM0
~~~

En `upload_port` debes indicar el puerto en el que tienes conectada la placa, en windows será algo como `COM3`, `COM7`... etc. En Linux `/dev/ttyUSB0` (328P, 2560) o `/dev/ttyACM0` (Samd21, 32u4) entre otros. Si solo tienes un dispositivo serial conectado puedes comentar esta variable (`;upload_port = /dev/ttyUSB0`) y dejar que PlatformIO lo reconozca por si solo, esto es útil cuando al conectar y desconectar la placa varias veces del USB el sistema constantemente nos cambia el nombre del puerto y tenemos que volver a configurarlo.

> Es necesario definir el tamaño del buffer Serial debido a que las tramas GPS con frecuencia superan el límite de 64bytes que el framework de Arduino establece por defecto. 128bytes son sufucientes. En el caso de Duino Uno y Duino Mega podemos añadir el macro en tiempo de compilación através del archivo _platformio.ini_, tal cual lo puedes ver en el ejemplo de arriba. Para el Zero es distinto y tendremos que editar un fichero del framework de Arduino: _{$HOME}/.platformio/packages/framework-arduinosam/cores/samd/RingBuffer.h_ En esta librería está definida la constante `SERIAL_BUFFER_SIZE`, pero no tiene un `#ifndef SERIAL_BUFFER_SIZE` para evitar que se duplique la declaración cuando lo añadimos al _platformio.ini_. La solución pasa por editar directamente este fichero e indicar un valor de 256. **OJO**, esto es un fichero del core de Arduino, por lo que si actualizas PlatformIO o las librerías de Arduino el fichero será sobreescrito y habrá que volver a realizar el cambio.

>En ocasiones se ha producido un error de compilación al utilizar la plataforma atmelavr con los Duino Mega y Duino Uno. El error en concreto dice: `internal compiler error: Segmentation fault`. La solución se encontró en esta issue de Platform.IO y pasa por definir unos flags en el archivo _platformio.ini_. Editamos _platformio.ini_ y en la configuración del mega y uno añadimos: `build_unflags = -flto`. Parece que es un bug de GCC con gcc-avr 4.9.2.

## Configuración de aplicación

La configuración se encuentra predefinida en el struct `Conf` del archivo Trackio.h. Este struct se inicializa por defecto en el método `Trackio::loadConf()` (Trackio.cpp) y se almacena en EEPROM. Por favor, revisa la documentación de `struct Conf` en Trackio.h para obtener detalles de cada configuración

Para editar la configuración debes realizar las modificaciones necesarias en el archivo `static-conf.h` donde se incluyen todas las configuraciones disponibles. No edites directamente los ficheros de la librería como main.cpp, Trackio.cpp o Trackio.h.

### Configuración de Pines

En _platformio.ini_ le decimos al PlatformIO qué compilador y método de upload debe utilizar para nuestra placa, en _static-conf.h_ establecemos la configuración de pines según el microcontrolador y pcb Rhomb.io que se use. Para ello al inicio del archivo (static-conf.h) existen unos macros para establecer la configuración de pines, donde deberás seleccionar una placa:
~~~C
// boards:
//#define DEIMOS_SOCKET1_MEGA2560
//#define DEIMOS_SOCKET2_MEGA2560
//#define HALLEY_SOCKET2_MEGA2560
//#define HALLEY_SOCKET2_UNO
//#define DEIMOS_SOCKET1_ZEROSAMD21
#define DEIMOS_SOCKET1_UNO
~~~

y un microcontrolador:
~~~C
// microcontroller:
//#define PINS_MEGA2560
//#define PINS_ZEROSAMD21D
#define PINS_UNO
~~~

Con esta configuración se establecen los pines necesarios para utilizar el módulo Sim868 de Rhomb.io junto con la PCB y microcontrolador seleccionados.

## Modos operacionles

TrackioUno dispone de varios modos operacionales que establecen como se comporta el dispositivo:

* **OP_STARTUP (0)**: El modo inicial en el que se establecen las configuraciónes básicas: se configura el Sim868, se habilitan redes GPRS, se muestra información del modem en consola... etc.
* **OP_TCP (1)**: Modo de funcionamiento bidireccional. TrackioUno habrirá un puerto TCP y lo mantendrá abierto por siempre y al mismo tiempo enviará tramas en el intervalo de tiempo indicado. El servidor podrá enviar comandos operacionales para cambiar la configuración del sistema y controlar el estado de las GPIOS. En este modo las funciones de Sleep (ahorro de energía) no estarán disponibles.
* **OP_AUTO (2)**: Funcionamiento estandar. El dispositivo enviará tramas al servidor en el tiempo establecido en `cfg.gpsInterval`. Entre cada envío se podrá indicar si se desea entrar en modo de bajo consumo (PWR_DOWN en el mcu) o muy bajo consumo (mcu + sim868)
* **OP_LOW (4)**: Modo de bajo consumo. Se deshabilita cualquier envío de información al servidor y tanto el mcu como el Sim868 entran en modo de bajo consumo. Este modo se habilita automáticamente cuando se llega a alguno de los valores indicados en `cfg.requiredVbat`, `cfg.requiredVin` y `cfg.requiredVsys5v`
* **OP_RST (5)**: Ejecuta un reset del sistema. A efectos se lanzará un while infinito lo cual hará saltar el watchdog del mcu y se volverá a iniciar la aplicación en modo OP_STARTUP

Para habilitar uno de estos modos de forma remota deberemos enviar un comando desde el servidor con el siguiente formato:

```c
// Habilitar el modo OP_TCP
#CFG|op:1$
// Forzar un reset del sistema
#CFG|op:5$
```

Puedes ver en el anterior listado el resto de números para cada modo.

> El modo operacional por defecto siempre debe ser OP_STARTUP, en la configuración de `cfg.primaryOpmode`, se establece el modo al que se cambiará una vez el startup finalice, este debe ser OP_AUTO o OP_TCP. No tendría sentido utilizar como primarios los modos OP_LOW y OP_RST.

## Puertos UART y debug

Trackio necesita dos puertos UART, uno para comunicación con el modem y otro para mostrar información de debug por consola. Siempre que sea posible intentaremos utilizar UART's por hardware, pero en el caso del Atmega328p, donde solo tenemos 1 puerto hardware, este lo usamos para el modem y creamos un puerto via SoftwareSerial para debug. En UART de debug solo hacemos uso del canal TX y en el caso de Atmega328p por defecto seleccionamos el pin _SDA_ como TX.

La configuración de los puertos UART se realiza en el archivo _rhio-pins.h_ en función de la PCB y módulo seleccionado en _static-conf.h_.

Para obtener un debug del dispositivo deberemos conectarnos, en el caso del atmega328p, a los pines _SDA_ y _GND_ con una aplicación Serial a 9600 baudios. Si usamos otro mcu podemos seguir el estandar de Rhomb.io, donde `Serial` se encuentra en los pines `A_TX` y `A_RX` y `Serial1` en `B_TX` y `B_RX`.

## Formato de tramas GPS

Breve resumen. La trama se ha divido en varios grupos de información

|   IMEI    | GPS DATA | BATTERY DATA |
|    :-:    |   :-:    |     :-:      |
| 12 cifras | AT+CGNSINF | vbat,vin,vsys |

> El IMEI solo se enviará cuando se seleccione el modo operacional OP_AUTO, OP_TCP no enviará IMEIs ya y que el servidor remoto almacena el IMEI de la sesión cuando se realiza la primera conexión (`Trackio::sayHello()`)

Se utiliza un formato de texto plano, cada grupo se separa con el símbolo `|`. Dentro de cada grupo los valores se separan con comas

GPS data contiene los siguientes valores:

* Tipo de trama, generalmente un 1, pero puede ser 0 si no hay información de geolocalización
* latitude
* longitude
* altura
* velocidad (sog)
* orientación (cog)
* [HDOP](https://en.wikipedia.org/wiki/Dilution_of_precision_(navigation))
* Nº de satélites GPS
* Calidad de la [señal GSM](http://m2msupport.net/m2msupport/atcsq-signal-quality/)

Cuando el dispositivo no dispone de cobertura GPS se seguirán enviando datos al servidor, el tipo de trama se reemplazará por 0

~~~
// trama sin GPS disponible
867717036798487|0,23|3929,36923,5023
~~~

> La info de batería siempre aparecerá representada aunque sea con ceros: `{IMEI}|...|3872,0,0`

>Las tramas se construyen en el método Trackio::transmitGps()

## Copyright

This software is property of M2M System Source S.L. Developed by Jordi Enguídanos with the contribution of Guillermo Alonso and Pedro Peláez


CHANGELOG.md
# Changelog

* **0.3.2 - 2019-01-21**
  * static-conf.h: ADD UART_BUFFER_SIZE para asignar tamaño de la variable buffer en Trackio.cpp
  * Trackio::sendCommand(): FIX bug overflow del buffer
  * Trackio::gprsIsOpen(): FIX, no se procesaba correctamente la respuesta de AT+CGATT?
  * main.cpp->transmitAlive(): FIX bug al reconectar el TCP, realizar registro en el gateway con Trackio::sayHello()
  * Trackio::applyConf(): ADD transmissionClock, requiredVbat/Vin/Vsys configurables de forma remota
  * Trackio::processCommand(): ADD comando "reset" realiza un hardReset -> #RST|1$
  * Trackio::processCommand(): ADD comando "save" guarda la configuración -> #SAVE|1$
  * Trackio::cmd_setConf(): eliminado el guardado automatica al recibir una nueva configuración remota
  * Trackio::hardReset(): ADD reset del Sim868 con "Trackio::powerOff()" previo al hardreset
  * Trackio::openTcp(): ADD reset de Trackio::transmissionClockCounter
  * Trackio::hardReset(): FIX uso correcto de hardReset()
  * Trackio.cpp/h: ADD soporte para lectura de baterías con HalleyBox
  * Trackio::enableGprs() renombrado como Trackio::openGprs()
  * Trackio::begin(): Cambio en secuencia de arranque para ejecutar ATE0
  * Minor Fixes

* **0.3.1 - 2018-11-15**
  * FIX bug al procesar comando CIPSTART, se evaluaba la linea 1 en lugar de la 2
  * Añadido parametro timeout a sendAt
  * Reconectar TCP si hay fallo al enviar alive
  * Trackio::sayHello(): Enviar el número de versión del firmware

* **0.3.0 - 2018-12-15**
  * Refactoring para Arduino Zero (veníamos de Atmega328p): Añadidas librerías FlashStorage y Adafruit Sleepydog y ReadLine
  * Nuevo método sendAt reemplaza sendCommand. Mejorada la comunicación con el modem
  * Gestión de acelerometro y alarma de movimiento
  * Otros arreglos y limpieza

* **0.2.6 - 2018-10-18**
  * FIX IO7, MOSI y MISO no estaban inicializadas como outputs

* **0.2.5 - 2018-10-16**
  * Todos los valores de configuración se han transferido a `static-conf.h` para evitar modificaciones "innecesarias" en archivos CORE.
  * Añadidas 3 nuevos GPIOS configurables de forma remota, IO7, MOSI y MISO. Eliminada LED01 como configurable de forma remota
  * FIX BUG. Se detecta un error al entrar en modo sleep (Trackio::sleepNow()) por el cual el dispositivo nunca llegaba a entrar en modo bajo consumo (OP_LOW) y esto podía impedir que la batería se cargara correctamente cuando el potenciomtro (P1) tuviera un nivel de carga bajo
  * FIX. Las variables requriedVbat/Vsys5v/Vin no estaban incluídas por defecto en la configuración inicial (Trackio::loadConf())
  * FIX. Trackio::openTcp() existía una comprobación redundante al verificar si el puerto estaba realmente abierto
  * Eliminadas algunas variables en desuso y ajustados los tipos de datos para reducir consumo de ram
  * Cambios menores.

* **0.2.4 - 2018-10-15**
  * En modo OP_TCP se utiliza el método `serialEvent()` (main.cpp) de Arduino para evitar el pool constante al `Serial.available()` en el loop principal.
  * En modo OP_TCP se añade un nuevo pool al servidor (`transmitAlive()`) y su variable de configuración `cfg.transmissionClock`. Este pool junto al anterior serialEvent() permiten que la transmisión de comandos servidor->cliente se realice de forma más eficiente.
  * Valores _required_ de vbat, vin y vsys incluídos en la configuración general
  * Actualizado README.md con información detallada
  * Los ACK al servidor siempre devuelven el estado de la IO6
  * Nuevo archivo de configuración static-conf.h
  * Se crean archivos con extensión .sample para static-conf.h y platformio.ini con el fin de evitar subir los archivos originales a los repositorios públicos.


* **0.2.3 - 2018-10-05**
  * Fix watchdog reset repentino en el startup, se añade wdt_disable() en el setup
  * Enviar en trama VSYS y VIN junto con VBAT
  * Enviar GSM junto con batería cuando no hay GPS
  * Eliminado cfg.battInterval, se mide batería siempre que se envía trama


* **0.2.2 - 2018-10-03**
  * Añadido powerOff en la secuencia de arranque
  * Añadido watchdog reset cuando openTcp falla


* **0.2.1 UNO - 2018-07-27**
  * Eliminado todo el código de Duino Zero, incluídas las librerías externas
    de RTC y SleepyDog para dar compatibilidad exclusiva a Duino Uno
  * Eliminadas variables de configuración y otras funciones (como bluetoth) para
    reducir consumo de memoria de ram.
  * Se utilizan las funciones de eeprom propias de AVR para guardar la configuración
    general. El objeto de configuración se saca de la clase trackio y se convierte
    en un struct global.
  * Añadida lectura analógica de baterías para VBAT, VIN y VSYS.


* **0.2.0 - 2018-07-12:**
  * OP_TCP: timer para verificar estado de la conexión tcp y reconectar en caso necesario
  * Duino Zero: Implementado watchdog+sleep con la librería SleepyDog
  * RTC: Implementado un reloj simple obteniendo la fecha desde servidor web. Se utiliza
    la librería timeLib.h para parsear el unix timestamp en fechas para humanos
  * Se vacía el setup() del main, OP_STARTUP lo reemplaza. Implementado mecanismo de
    protección ante fallos para reestablecer modo OP_STARTUP si necesario
  * OP_AUTO abre/cierra el TCP cuando necesario
  * Tramas de datos: Datos de GPS, Sensing y CAN unificados en una unica trama
    y agregado el método transmitData() para realizar el envío
  * Método sendCommand, agregada limpieza de buffer y aumentado de 15 a 25 el número
    de intentos para esperar respuesta


lib/TLA2024/TLA2024.h
#include <Arduino.h>
#include <Wire.h>

#pragma once
#define SINGLE true
#define CONT false

#define AIN0	4
#define AIN1	5
#define AIN2	6
#define AIN3	7

class TLA2024 {
 public:
  TLA2024();

  /*
    Initializes I2C bus
    returns:
       0 - adc responds with correct default conf
       1 - otherwise
  */
  int begin();

  // resets device to default configuration
  void reset();

  // restores device to last config
  void restore();

  float analogRead();

  /*
    Sets the Full Scale Range of the ADC

    gain:
      60 -> ± 6.144v
      40 -> ± 4.096
      20 -> ± 2.048 (default)
      10 -> ± 1.024
      5  -> ± 0.512
      2  -> ± 0.256
  */
  void setFSR(uint8_t gain);

  /*
    Configures the input signals to the ADC

    option:
      1 -> P = 0, N = 1 (TP1 - TP2)
      2 -> P = 0, N = 3 (TP1 - VGND)
      3 -> P = 1, N = 3 (TP2 - VGND)
      4 -> P = 0, N = GND
      5 -> P = 1, N = GND
      6 -> P = 2, N = GND
      7 -> P = 3, N = GND
  */
  void setMux(uint8_t option);

  /*
    Continous conversion (0) or single shot (1)
  */
  void setMode(bool mode);

  void setDR(uint8_t rate);

 private:
  uint8_t addr = 0x48;
  uint8_t conv_reg = 0x00;
  uint8_t conf_reg = 0x01;

  // this is default conf.
  uint16_t init_conf = 0x8583;
  uint16_t saved_conf = 0x8583;
  bool mode = SINGLE;

  union I2C_data {
    uint8_t packet[2];
    uint16_t value;
  } data;

  /*
    A generic read from mem_addr.
  */
  uint16_t read(uint8_t mem_addr);

  /*
    We only write to the configuration register.
    out_data is the 16 bits of conf_regs

    Should always return 2

    Saves data to current_conf
  */
  int write(uint16_t data);
};

TLA2024::TLA2024() {}

int TLA2024::begin() {
  Wire.begin();
  Wire.setClock(100000);
  uint16_t init = read(conf_reg);
  // make sure communication with device is working and that it is OK
  if ((init == init_conf) || 1) {
    return 0;
  } else {
    return 1;
  }
}

uint16_t TLA2024::read(uint8_t mem_addr) {
  Wire.beginTransmission(addr);
  Wire.write(mem_addr);
  Wire.endTransmission();
  delay(5);
  Wire.requestFrom(addr, 2);
  if (2 <= Wire.available()) {
    // bring in data
    data.packet[1] = Wire.read();
    data.packet[0] = Wire.read();
    uint16_t ret = data.value;
    data.value = 0;
    return ret;
  }
  return -1;
}

int TLA2024::write(uint16_t out_data) {
  int written = 0;
  // save conf
  saved_conf = out_data;
  // put our out_data into the I2C data union so we can send MSB and LSB
  data.value = out_data;
  Wire.beginTransmission(addr);
  Wire.write(conf_reg);
  written += Wire.write(data.packet[1]);
  written += Wire.write(data.packet[0]);
  Wire.endTransmission();
  data.value = 0;
  return written;
}

void TLA2024::reset(void) { write(init_conf); }

void TLA2024::restore(void) {
  uint16_t restore_conf = saved_conf & ~0x8000;
  write(restore_conf);
}


float TLA2024::analogRead(void) {

  // this only needs to run when in single shot.
  if (mode) {
    // write 1 to OS bit to start conv
    uint16_t current_conf = read(conf_reg);
    current_conf |= 0x8000;
    write(current_conf);
    // OS bit will be 0 until conv is done.
    do {
      delay(5);
    } while ((read(conf_reg) & 0x8000) == 0);
  }

  // get data from conv_reg
  uint16_t in_data = read(conv_reg);

  // shift out unused bits
  in_data >>= 4;

  // get sign and mask accordingly
  if (in_data & (1 << 11)) {
    // 11th bit is sign bit. if its set, set bits 15-12
    in_data |= 0xF000;
  } else {
    // not set, clear bits 15-12
    in_data &= ~0xF000;
  }

  // now store it as a signed 16 bit int.
  int16_t ret = in_data;

  // default Full Scale Range is -2.048V to 2.047V.
  // our 12bit 2's complement goes from -2048 to 2047 :)
  // return ret /1000.0;

  // return raw adc data
  return ret;
}


void TLA2024::setFSR(uint8_t gain) {
  // bring in conf reg
  uint16_t conf = read(conf_reg);
  // clear the PGA bits:
  conf &= ~0x0E00;

  switch (gain) {
    case 60:
      // PGA bits already cleared
      break;

    case 40:
      // set bit 9
      conf |= 0x0200;
      break;

    case 20:
      // set bit 10
      conf |= 0x0400;
      break;

    case 10:
      // set bit 10-9
      conf |= 0x0600;
      break;

    case 5:
      // set bit 11
      conf |= 0x0800;
      break;

    case 2:
      // set bit 11-9
      conf |= 0x0E00;
      break;
  }
  write(conf);
}

void TLA2024::setMux(uint8_t option) {
  // bring in conf reg
  uint16_t conf = read(conf_reg);
  // clear MUX bits
  conf &= ~0x7000;

  switch (option) {
    case 1:
      // bits already cleared
      break;

    case 2:
      // set bit 12
      conf |= 0x1000;
      break;

    case 3:
      // set bit 13
      conf |= 0x2000;
      break;

    case 4:
      // set bit 14
      conf |= 0x4000;
      break;

    case 5:
      // set bits 14 and 12
      conf |= 0x5000;
      break;

    case 6:
      // set bits 14, and 13
      conf |= 0x6000;
      break;

    case 7:
      // set bits 14, 13 and 12
      conf |= 0x7000;
      break;
  }
  write(conf);
}

void TLA2024::setMode(bool mode) {
  // bring in conf reg
  TLA2024::mode = mode;
  uint16_t conf = read(conf_reg);
  // clear MODE bit (8) (continous conv)
  conf &= ~(1 << 8);
  if (mode) {
    // single shot
    conf |= (1 << 8);
  }
  write(conf);
}

void TLA2024::setDR(uint8_t rate) {
  // bring in conf reg
  uint16_t conf = read(conf_reg);
  // set bits 7:5
  conf |= 0b111 << 5;
  write(conf);
}


lib/Trackio/rhio-pins.h
#ifndef TRACKIO_PINS
#define TRACKIO_PINS

#ifdef DEIMOS_SOCKET1_MEGA2560
  #define GSM_PWREN  36 // IO00
  #define GSM_PWRKEY 33  // IO01
  #define GPS_EN     44 // IO02
  #define GSM_STATUS 37 // IO03

  #define SerialSim Serial // comunicación con SIMCOM
  #define SerialMon Serial1 // debug, terminal
#endif

#ifdef DEIMOS_SOCKET2_MEGA2560
  #define GSM_PWREN  47 // IO08
  #define GSM_PWRKEY 42  // IO09
  #define GPS_EN     45 // IO10
  #define GSM_STATUS 4 // IO11

  #define SerialSim Serial1 // comunicación con SIMCOM
  #define SerialMon Serial // debug, terminal
#endif

#ifdef HALLEY_SOCKET2_MEGA2560
  #define GSM_PWREN  47 // IO08
  #define GSM_PWRKEY 42  // IO09
  #define GPS_EN     45 // IO10
  #define GSM_STATUS 4 // IO11

  #define SerialSim Serial1 // comunicación con SIMCOM
  #define SerialMon Serial // comunicación con SIMCOM
#endif

#ifdef HALLEY_SOCKET2_UNO
  #define GSM_PWREN  8
  #define GSM_PWRKEY 9
  #define GPS_EN     10
  #define GSM_STATUS 11

  #define SerialSim Serial     // comunicación con SIMCOM
  extern SoftwareSerial SoftSerial;
  #define SerialMon SoftSerial // comunicación con SIMCOM
#endif

#ifdef DEIMOS_SOCKET1_ZEROSAMD21
  #define GSM_PWREN  2
  #define GSM_PWRKEY 5
  #define GPS_EN     6
  #define GSM_STATUS 7

  #define SerialSim Serial1    // comunicación con SIMCOM
  #define SerialMon Serial // comunicación con SIMCOM
  // #define SerialMon SerialUSB // comunicación con SIMCOM
#endif

#ifdef DEIMOS_SOCKET1_UNO
  #define GSM_PWREN  4
  #define GSM_PWRKEY 5
  #define GPS_EN     6
  #define GSM_STATUS 7

  #define SerialSim Serial     // comunicación con SIMCOM
  extern SoftwareSerial SoftSerial;
  #define SerialMon SoftSerial // comunicación con SIMCOM
#endif

#ifdef PINS_MEGA2560
  #define IO00 36
  #define IO01 33
  #define IO02 44
  #define IO03 37
  #define IO04 34
  #define IO05 43
  #define IO06 46
  #define IO07 48
  #define IO08 47
  #define IO09 42
  #define IO10 45
  #define IO11 4
  #define IO12 32
  #define IO13 26
  #define IO14 31
  #define IO15 27
  #define IO16 39
  #define IO17 25
  #define IO18 24
  #define IO19 30
  #define IO20 28
  #define IO21 23
  #define IO22 29
  #define IO23 22
  #define IO24 35
  #define IO25 40
  #define IO26 41
  #define IO27 38
  #define IO28 28
  #define IO29 9
  #define IO30 8
  #define IO31 7
  #define LED01 13
#endif

#ifdef PINS_ZEROSAMD21D
  #define NMI    4
  #define INT0   3
  #define INT1   23
  // #define CAPT0   ??   // PA28, pin not includded on Arduino IDE for original Zero

  #define IO0    2
  #define IO1    5
  #define IO2    6
  #define IO3    7
  #define IO4    8
  #define IO5    -1   // PA27, pin not includded on Arduino IDE for original Zero
  #define IO6    24
  #define IO7    22

  #define LED    9
  #define PWM0   9
  // #define PWM1   ??  // PB03, pin not includded on Arduino IDE for original Zero

  #define RXD    0
  #define TXD    1
  // #define RTS    ??  // PA31, pin not includded on Arduino IDE for original Zero
  // #define CTS    ??  // PA30, pin not includded on Arduino IDE for original Zero
  #define RXD1   31
  #define TXD1   30

  #define SDA    20
  #define SCL    21
  #define MOSI   35
  #define MISO   34
  #define SCK    37
  #define CS0    10
  #define CS1    A5  // MUX_SW=LOW required
  #define MEM_CS A4  // MUX_SW=LOW required
  #define uSD_CS A2  // MUX_SW=LOW required
  // #define USB_P      ??  // PA25, pin not includded on Arduino IDE for original Zero
  // #define USB_N     ??  // PA24, pin not includded on Arduino IDE for original Zero
  #define _1WIRE A3  // MUX_SW=LOW required

  #define AD0    A1
  #define AD1    A4  // MUX_SW=HIGH required
  #define AD2    A2  // MUX_SW=HIGH required
  #define AD3    A5  // MUX_SW=HIGH required
  #define DAC0   A0

  #define MUX_SW 38  // ANALOG (MULTIPLEXOR) SWITCH
#define RSTOUT A3 // MUX_SW=HIGH required
#endif

#ifdef PINS_UNO
  #define NMI 2
  #define _INT0 3
  #define IO0 4
  #define IO1 5
  #define IO2 6
  #define IO3 7
  #define IO4 8
  #define IO5 A0
  #define IO6 A1
  #define IO7 A2
  #define LED01 9
  #define PWM0 9
  #define RXD 0
  #define TXD 1
  #define SDA A4
  #define SCL A5
  #define MOSI 11
  #define MISO 12
  #define SCK 13
  #define CS0 10
  #define _1WIRE A3 // SJ4 must be closed & SJ1 must be open
  #define AD0 A6
  #define CAPT0 A7
  #define RSTOUT A3 // SJ1 must be closed & SJ4 must be open
#endif

#endif // #ifndef TRACKIO_PINS
